File: leetcode/graph/1245_Tree_Diameter/codes/dfs.py
from typing import List
from collections import defaultdict

class Solution:
    def treeDiameter(self, edges: List[List[int]]) -> int:
        if not edges:
            return 0
        
        n = len(edges) + 1
        graph = defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        def dfs(node, parent, dist):
            farthest_node, max_dist = node, dist
            for nxt in graph[node]:
                if nxt != parent:
                    cand_node, cand_dist = dfs(nxt, node, dist + 1)
                    if cand_dist > max_dist:
                        farthest_node, max_dist = cand_node, cand_dist
            return farthest_node, max_dist
        
        node_b, _ = dfs(0, -1, 0)
        _, diameter = dfs(node_b, -1, 0)
        return diameter

File: leetcode/graph/1245_Tree_Diameter/codes/dfs.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *neighbors;
    int size;
    int capacity;
} AdjList;

void addEdge(AdjList* graph, int u, int v) {
    if (graph[u].size == graph[u].capacity) {
        graph[u].capacity *= 2;
        graph[u].neighbors = (int*)realloc(graph[u].neighbors, sizeof(int)*graph[u].capacity);
    }
    graph[u].neighbors[graph[u].size++] = v;
}

void dfs(int node, int parent, int dist, AdjList* graph, int *maxDist, int *farthestNode) {
    if (dist > *maxDist) {
        *maxDist = dist;
        *farthestNode = node;
    }
    for (int i = 0; i < graph[node].size; i++) {
        int nxt = graph[node].neighbors[i];
        if (nxt != parent) {
            dfs(nxt, node, dist + 1, graph, maxDist, farthestNode);
        }
    }
}

int treeDiameter(int** edges, int edgesSize, int* edgesColSize){
    int n = edgesSize + 1;
    AdjList* graph = (AdjList*)malloc(sizeof(AdjList)*n);
    for (int i = 0; i < n; i++) {
        graph[i].neighbors = (int*)malloc(sizeof(int)*2);
        graph[i].size = 0;
        graph[i].capacity = 2;
    }
    
    for (int i = 0; i < edgesSize; i++) {
        int u = edges[i][0], v = edges[i][1];
        addEdge(graph, u, v);
        addEdge(graph, v, u);
    }
    
    int maxDist = -1, farthestNode = 0;
    dfs(0, -1, 0, graph, &maxDist, &farthestNode);
    
    maxDist = -1;
    dfs(farthestNode, -1, 0, graph, &maxDist, &farthestNode);
    
    for (int i = 0; i < n; i++) free(graph[i].neighbors);
    free(graph);
    
    return maxDist;
}
